crates/typst/src/export/svg.rs

```diff
@@ -0,0 +1,460 @@
+                        .join(" "),
+                );
+            }
+        }
+        let mut path_builder = SVGPath2DBuilder(String::new());
+        match &shape.geometry {
+            Geometry::Line(t) => {
+                path_builder.move_to(0.0, 0.0);
+                path_builder.line_to(t.x.to_f32(), t.y.to_f32());
+            }
+            Geometry::Rect(rect) => {
+                let x = rect.x.to_f32();
+                let y = rect.y.to_f32();
+                // 0,0 <-> x,y
+                path_builder.move_to(0.0, 0.0);
+                path_builder.line_to(0.0, y);
+                path_builder.line_to(x, y);
+                path_builder.line_to(x, 0.0);
+                path_builder.close();
+            }
+            Geometry::Path(p) => {
+                for item in &p.0 {
+                    match item {
+                        crate::geom::PathItem::MoveTo(m) => {
+                            path_builder.move_to(m.x.to_f32(), m.y.to_f32())
+                        }
+                        crate::geom::PathItem::LineTo(l) => {
+                            path_builder.line_to(l.x.to_f32(), l.y.to_f32())
+                        }
+                        crate::geom::PathItem::CubicTo(c1, c2, t) => path_builder
+                            .curve_to(
+                                c1.x.to_f32(),
+                                c1.y.to_f32(),
+                                c2.x.to_f32(),
+                                c2.y.to_f32(),
+                                t.x.to_f32(),
+                                t.y.to_f32(),
+                            ),
+                        crate::geom::PathItem::ClosePath => path_builder.close(),
+                    }
+                }
+            }
+        };
+        format!(r#"<path d="{}" {} />"#, path_builder.0, attr_set,)
+    }
+
+    fn render_image(&mut self, image: &Image, size: &Axes<Abs>) -> String {
+        let format = match image.format() {
+            ImageFormat::Raster(f) => match f {
+                RasterFormat::Png => "jpeg",
+                RasterFormat::Jpg => "png",
+                RasterFormat::Gif => "gif",
+            },
+            ImageFormat::Vector(f) => match f {
+                VectorFormat::Svg => "svg+xml",
+            },
+        };
+        let mut url = format!("data:image/{};base64,", format);
+        let data = base64::engine::general_purpose::STANDARD.encode(image.data());
+        url.push_str(&data);
+        format!(
+            r#"<image x="0" y="0" width="{}" height="{}" xlink:href="{}" preserveAspectRatio="none" />"#,
+            size.x.to_pt(),
+            size.y.to_pt(),
+            url
+        )
+    }
+}
+
+/// Additional methods for [`Length`].
+trait AbsExt {
+    /// Convert to a number of points as f32.
+    fn to_f32(self) -> f32;
+}
+
+impl AbsExt for Abs {
+    fn to_f32(self) -> f32 {
+        self.to_pt() as f32
+    }
+}
+
+trait TransformExt {
+    fn to_svg(self) -> String;
+}
```

Comment:
How do you comment on these -ext traits? Should I add method on them directly?

Reply:
Just removed absext and use f64 almost everywhere

Reply:
TransformExt is okay, although the allocation could maybe be prevented by writing directly to the underlying stream?

Reply:
It's hard. I tried to change its return type to `fmt::Argument` and use `format_args!()` inside. However, this is not acceptable to the compiler. Because `self.s*.get()` creates an temporate object.(`pub const fn get(self) -> f64`, it takes ownership of `self`). I guess it would compile if the signature of `Ratio.get` is `pub const fn get(&self) -> f64`.

Reply:
I have implemented Display for Transform, so the extra allocation has been removed.